### PR TITLE
[node config] allow initial set of Move modules to be chosen at genesis

### PR DESCRIPTION
**Pull request by sblackshear**
The set of Move modules published at genesis is currently hardcoded. This makes it difficult to reuse the Diem software in a setting that requires a different set of initial Move modules or a different genesis state.

This PR adds the bytecode for initial Move modules (represented as list of hex strings) to the `layout` part of the genesis builder, then creates the genesis transaction using this list.

### CODE REVIEW

**Comment by sblackshear** [comment]
@mgorven I am trying to figure out the best way to incorporate setting the modules in a genesis ceremony (like the one you shared with me [here](https://github.com/diem/partners/blob/master/testnet/testnet/templates/genesis.yaml#L83); advice on that would be appreciated! 

One approach I considered was adding something like `diem-genesis-tool set_modules <hex_encoded_bytecodes>`, but that doesn't seem great considering how big `hex_encoded_bytecodes` will typically be (e.g., [here's](https://gist.github.com/sblackshear/0d6c9b4f9cf46661f200ead5677eb413) what it looks like for the current Diem Framework).

**Comment by davidiw** [comment]
I think one approach would be to have a folder called move or move_modules and within their have your encoded modules. The layout would then contain just the names of those files. The intent being that github manages versions, so that one need only use the correct git commit from the genesis repo to generate genesis.blob.

**Comment by sblackshear** [comment]
> I think one approach would be to have a folder called move or move_modules and within their have your encoded modules. The layout would then contain just the names of those files. The intent being that github manages versions, so that one need only use the correct git commit from the genesis repo to generate genesis.blob.

I would certainly prefer passing around a filename instead of the giant list of hex strings... But it appeared to me that the genesis config builder code was intentionally trying to avoid reading from files/directories that it did not create, so I thought that there might be some constraint on using external directories like `move_modules` to build the configs.

**Comment by davidiw** [comment]
> I would certainly prefer passing around a filename instead of the giant list of hex strings... But it appeared to me that the genesis config builder code was intentionally trying to avoid reading from files/directories that it did not create, so I thought that there might be some constraint on using external directories like move_modules to build the configs.

Not at all, the layout file literally is a pointer to other folders that contain additional configuration (public keys). I do know that internally it all eventually gets consumed by something that wants everything in memory but not at the CLI layer more at the library layer.

**Comment by bmwill** [comment]
> Not at all, the layout file literally is a pointer to other folders that contain additional configuration (public keys). I do know that internally it all eventually gets consumed by something that wants everything in memory but not at the CLI layer more at the library layer.

This isn't true for all secure backends. It's true that the Layout includes "pointers" to where other data lives but those "pointers" are not directories or files on disk but rather other namespaces in the same secure storage. For an OnDisk Storage implementation this does translate into directories or files on disk but this isn't the case for a Vault or InMemory Storage impl.

So while yes it would be ok at the CLI layer to supply a path to the modules that should be included in genesis, the actual Layout used shouldn't specify files on disk but rather some other place in the secure storage where the data lives. The CLI tool could read those files and then store them in the secure store

**Comment by sblackshear** [comment]
Got it, thanks for the clarifications! I will include a pointer into a secure storage namespace where module bytecodes live instead + add CLI logic for loading the bytecodes into secure storage.

**Comment by sblackshear** [comment]
Done. I think this looks much better than the hex strings.

File: config/management/genesis/src/command.rs [hunk 1]
```diff
@@ -213,7 +223,23 @@ pub mod tests {
...
 
-        // Step 2) Upload the root keys:
+        // Step 2) Upload the Move modules
+        let tempdir = diem_temppath::TempPath::new();
+        tempdir.create_as_dir().unwrap();
+        for b in diem_framework_releases::current_module_blobs() {
+            let mut temppath =
+                diem_temppath::TempPath::new_with_temp_dir(tempdir.path().to_path_buf());
+            temppath.create_as_file().unwrap();
+            temppath.persist(); // otherwise, file will disappear before we call set_move_modules
```
Comment (davidiw) [inline config/management/genesis/src/command.rs:233]: I imagine this gets cleaned up by the tempdir?

**Review by davidiw — COMMENTED** [review]
could you maybe update the documentation with the new process? https://github.com/diem/diem/blob/main/config/management/genesis/README.md

should we add to the layout the location of the move-modules? or is it expected to be implicit?

File: config/management/genesis/src/command.rs (hunk 1 above)
Comment (sblackshear) [reply config/management/genesis/src/command.rs:233]: Correct.

**Comment by sblackshear** [comment]
> should we add to the layout the location of the move-modules? or is it expected to be implicit?

My understanding of the layout (based on reading the docs/code, though I could be wrong) is that it encodes the names and roles of the participants in the system (e.g., DR, TC, validators). Since the initial Move modules aren't participants, I thought it wasn't appropriate to include them under the layout.

**Review by davidiw — APPROVED** [review]
cool stuff!

**Comment by zihaoccc** [comment]
```
2021-07-15T15:00:35.365Z [ERROR] core: no seal config found, can't determine if legacy or new-style shamir
2021-07-15T15:00:32.900081Z [main] ERROR common/crash-handler/src/lib.rs:38 details = '''panicked at 'Storage is not available: InternalError("Synthetic error returned: Connection Failed: Connection refused (os error 111)\n")', consensus/safety-rules/src/safety_rules_manager.rs:22:9'''
```
seems something wrong with vault and lsr pods

**Comment by sblackshear** [comment]
Is there a way to find out which test this is coming from? I'd like to look at the terraform configs and see if it's due to a missing `--set-move-modules` that's now required after this PR.

**Comment by zihaoccc** [comment]
> Is there a way to find out which test this is coming from? I'd like to look at the terraform configs and see if it's due to a missing `--set-move-modules` that's now required after this PR.

This is coming from LBT test AKA compatibility test. reference to log: http://kibana.ct-1-k8s-testnet.aws.hlw3truzy4ls.com/app/kibana#/discover?_g=(time:(from:'2021-07-15T14:58:41Z',to:'2021-07-15T15:00:45Z'))&_a=(columns:!(log),index:'1ffd73f0-4182-11eb-ac99-f5a7597136e1',interval:auto,query:(language:kuery,query:'kubernetes.pod_name:%5C%22val-1%5C%22'),sort:!(!('@timestamp',desc)))

**Comment by zihaoccc** [comment]
> Is there a way to find out which test this is coming from? I'd like to look at the terraform configs and see if it's due to a missing `--set-move-modules` that's now required after this PR.

found 
```
You may need to set the following environment variable:
```
if you change requires new env value, I think u should have import it from somewhere otherwise comp test expects a failed

**Comment by sblackshear** [comment]
> You may need to set the following environment variable:

Ah, good sleuthing, thank you! Does it say which environment value (is it something like `move-modules` or `MOVE_MODULES`)?

> if your change requires new env value, I think u should have to import it from somewhere otherwise comp test expects a failed

Do you know how I can set up this import?

**Comment by zihaoccc** [comment]
> sleuthing

It did not say, seems like the name of the env variable has not been lodged into log.
From my perspective, I think LBT currently rely on CT backend k8s which dose not require deployment setup change for passing in env by terraform, grab @rustielin for confirming. Maybe u can pass in the value in fn
```
let args = format!(
            "
                diem-genesis-tool
                set-move-modules
                --dir {dir}
                --shared-backend backend={backend};\
                    path={storage_path};\
                    namespace={ns}
            ",
            dir = dir,
            backend = DISK,
            storage_path = self.path,
            ns = namespace,
        );
```
?

**Comment by zihaoccc** [comment]
> > You may need to set the following environment variable:
> 
> Ah, good sleuthing, thank you! Does it say which environment value (is it something like `move-modules` or `MOVE_MODULES`)?
> 
> > if your change requires new env value, I think u should have to import it from somewhere otherwise comp test expects a failed
> 
> Do you know how I can set up this import?

It did not say, seems like the name of the env variable has not been lodged into log.
From my perspective, I think LBT currently rely on CT backend k8s which dose not require deployment setup change for passing in env by terraform, grab @rustielin for confirming. Maybe u can pass in the value in fn

```
let args = format!(
            "
                diem-genesis-tool
                set-move-modules
                --dir {dir}
                --shared-backend backend={backend};\
                    path={storage_path};\
                    namespace={ns}
            ",
            dir = dir,
            backend = DISK,
            storage_path = self.path,
            ns = namespace,
        );
```
?

**Comment by sblackshear** [comment]
> > > You may need to set the following environment variable:
> > 
> > 
> > Ah, good sleuthing, thank you! Does it say which environment value (is it something like `move-modules` or `MOVE_MODULES`)?
> > > if your change requires new env value, I think u should have to import it from somewhere otherwise comp test expects a failed
> > 
> > 
> > Do you know how I can set up this import?
> 
> It did not say, seems like the name of the env variable has not been lodged into log.
> From my perspective, I think LBT currently rely on CT backend k8s which dose not require deployment setup change for passing in env by terraform, grab @rustielin for confirming. Maybe u can pass in the value in fn
> 
> ```
> let args = format!(
>             "
>                 diem-genesis-tool
>                 set-move-modules
>                 --dir {dir}
>                 --shared-backend backend={backend};\
>                     path={storage_path};\
>                     namespace={ns}
>             ",
>             dir = dir,
>             backend = DISK,
>             storage_path = self.path,
>             ns = namespace,
>         );
> ```
> 
> ?

Interesting... I copied that code from

```
  pub async fn set_layout(&self, path: &str, namespace: &str) -> Result<Layout, Error> {
        let args = format!(
            "
                diem-genesis-tool
                set-layout
                --path {path}
                --shared-backend backend={backend};\
                    path={storage_path};\
                    namespace={ns}
            ",
            path = path,
            backend = DISK,
            storage_path = self.path,
            ns = namespace,
        );
        let command = Command::from_iter(args.split_whitespace());
        spawn_blocking(|| command.set_layout())
            .await
            .expect("tokio spawn_blocking runtime error")
    }
```
in genesis-helper which seems to work fine...

File: config/management/genesis/README.md [hunk 2]
```diff
@@ -4,16 +4,17 @@ The `diem-genesis-tool` provides a tool for the genesis ceremony of the Diem blo
 
 * A diem root account that maintains the set of validator owners, validator operators, and the active validator set.
 * A treasury compliance account that maintains VASPs, DDs, and other related topics.
+* The initial set of Move modules published at genesis.
 * Validator owners (OW) that have accounts on the blockchain. These accounts contain a validator configuration and specify a validator operator.
 * Validator operators (OP) that have accounts on the blockchain. These accounts have the ability to manipulate validator configuration.
-
```
Comment (davidiw) [inline config/management/genesis/README.md:9]: do you hate whitespace 😡

**Review by davidiw — APPROVED** [review]
/land

File: config/management/genesis/README.md (hunk 2 above)
Comment (sblackshear) [reply config/management/genesis/README.md:9]: so much

**Pull request closed without merging**